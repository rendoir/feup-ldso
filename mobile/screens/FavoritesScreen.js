import React from 'react';
import {
    ScrollView,
    StyleSheet,
    RefreshControl
} from 'react-native';
import { Font, AppLoading } from "expo";
import { Root, View, Card, Icon, Text } from 'native-base';
import axios from 'axios';
import { SecureStore } from 'expo';
import FavoriteEvent from '../components/FavoriteEvent';
import CustomHeader from '../components/CustomHeader';

export default class FavoritesScreen extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            loading: true,
            events: [],
            updateCall: false,
            eventsPage: 0,
            refreshing: false
        };

        this.handleScroll = this.handleScroll.bind(this);
        this.onFavorite = this.onFavorite.bind(this);
        this.getFavoriteEvents = this.getFavoriteEvents.bind(this);
        this._onRefresh = this._onRefresh.bind(this);
    }

    static navigationOptions = {
        header: null
    };

    componentWillUnmount() {
        this.isCancelled = true;
    }

    async componentDidMount() {
        await Font.loadAsync({
            Roboto: require("native-base/Fonts/Roboto.ttf"),
            Roboto_medium: require("native-base/Fonts/Roboto_medium.ttf")
        });
        await Font.loadAsync({
            'OpenSans-Regular': require('../assets/fonts/OpenSans-Regular.ttf'),
            'DJB-Coffee-Shoppe-Espresso': require('../assets/fonts/DJB-Coffee-Shoppe-Espresso.ttf')
        });
        this.getFavoriteEvents();
        this.setState({ loading: false, events: [], eventsPage: 0 });
        this.didFocusSubscription();
    }

    async getFavoriteEvents() {
        let token = await SecureStore.getItemAsync('access_token');
        let self = this;
        let apiLink = 'http://' + global.api + ':3030/events/favorites?limit=' + 10 + '&offset=' + (this.state.eventsPage * 10) + '&';
        apiLink += "user_id=" + global.userId + '&';
        apiLink += "token=" + token + '&';
        axios.get(apiLink)
            .then(function(response) {

                if (!this.isCancelled) {
                    const evs = response.data.map((event) => {
                        let isFav = false;
                        if (event.is_favorite !== undefined)
                            isFav = event.is_favorite;
                        else if (event.favorite !== undefined)
                            isFav = (event.favorite.length == 1);

                        return { ...event, is_favorite: isFav };
                    });
                    self.setState({ events: [...self.state.events, ...evs] });
                }

            })
            .catch(function(error) {
                console.log(error);
            });
        this.setState({ updateCall: false });
    }

    onSelect = updateCall => {
        this.setState(updateCall);
    };

    didFocusSubscription() {
        this.props.navigation.addListener('didFocus', () => {
            this.setState({ events: [], eventsPage: 0 });
            this.getFavoriteEvents();
        });
    }

    async onFavorite(event_id) {
        let token = await SecureStore.getItemAsync('access_token');
        let self = this;

        let apiLink = 'http://' + global.api + ':3030/favorite';
        axios.post(apiLink,
            {
                user_id: global.userId,
                event_id: event_id,
                token: token
            })
            .then(function() {
                const list = self.state.events.map((ev) => (
                    ev.id == event_id ? { ...ev, is_favorite: !ev.is_favorite } : ev
                ));
                self.setState({ events: list });
            })
            .catch(function(error) {
                console.log(error);
            });
    }

    handleScroll(event) {
        let scrollOffsetY = event.nativeEvent.contentOffset.y;
        if (scrollOffsetY > (5 * (this.state.eventsPage + 1) * 130)) {
            let page = this.state.eventsPage + 1;
            this.setState({ eventsPage: page });
            this.getFavoriteEvents();
        }
    }

    _onRefresh() {
        this.setState({ refreshing: true, events: [], eventsPage: 0 });
        this.getFavoriteEvents();
        this.setState({refreshing: false});
    }

    render() {
        if (this.state.loading) {
            return (
                <Root>
                    <AppLoading />
                </Root>
            );
        }
        const { navigate } = this.props.navigation;

        const events = this.state.events.map((event) => (
            <FavoriteEvent language={this.props.screenProps.language} {...event} key={event.id} onPress={() => navigate('Event', { eventData: event })} onFavorite={this.onFavorite} />
        ));

        if (this.state.updateCall)
            this.getFavoriteEvents();

        let noEventsElement;
        if (this.state.events.length == 0) {
            noEventsElement = (
                <Card>
                    <View style={{ justifyContent: 'center', alignItems: 'center', marginVertical: '5%' }}>
                        <Icon type="Feather" name="info" />
                        <Text>{global.dictionary["NO_FAVORITES"][this.props.screenProps.language]}</Text>
                    </View>
                </Card>
            );
        }

        return (
            <View style={styles.wrapperView}>
                <CustomHeader language={this.props.screenProps.language} toggleLanguage={this.props.screenProps.toggleLanguage}/>
                <ScrollView className="scroll_view" refreshControl={ <RefreshControl refreshing={this.state.refreshing} onRefresh={this._onRefresh} /> } style={styles.scrollView} onScroll={this.handleScroll}>

                    <View style={styles.favoritesLabelView}>
                        <Text style={styles.favoritesLabelText}>{global.dictionary["FAVORITES"][this.props.screenProps.language]}</Text>
                    </View>

                    <View style={styles.eventsView}>
                        {events}
                        {noEventsElement}
                    </View>
                </ScrollView >
            </View >
        );
    }
}

const styles = StyleSheet.create({
    eventsView: {
        marginHorizontal: '2%',
        backgroundColor: '#7C8589',
        marginBottom: '10%'
    },
    wrapperView: {
        backgroundColor: '#7C8589'
    },
    favoritesLabelText: {
        fontSize: 32,
        color: 'white',
        textAlign: 'center',
        fontFamily: 'OpenSans-Regular'
    },
    favoritesLabelView: {
        marginVertical: '3%',
        paddingHorizontal: '5%',
        backgroundColor: '#7C8589',
        justifyContent: 'center',
        alignItems: 'center'
    },
    scrollView: {
        backgroundColor: '#7C8589',
        height: '100%',
        marginBottom: '10%'
    }
});
