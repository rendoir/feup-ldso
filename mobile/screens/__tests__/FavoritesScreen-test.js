/**
* @jest-environment jsdom
*/
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { shallow } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FavoritesScreen from '../FavoritesScreen';

Enzyme.configure({ adapter: new Adapter() });

var mockAxios = new MockAdapter(axios);

describe('App snapshot', () => {
    jest.useFakeTimers();
    beforeEach(() => {
        NavigationTestUtils.resetInternalState();
    });

    it('renders correctly', async() => {
        const wrapper = shallow(<FavoritesScreen />);
        expect(wrapper).toMatchSnapshot();
    });

    it('mounts correctly', () => {
        const mockFn = jest.fn('didFocusSubscription');
        let i = 0;
        if (i == 1)
            mockFn();
        const tree = renderer.create(<FavoritesScreen />).getInstance();
        setImmediate(() => {
            tree.componentDidMount();
            expect(tree.state.isCancelled).toEqual(false);
            expect(tree.state.loading).toEqual(false);
        });
    });

    it('unmounts correctly', () => {
        const mockFn = jest.fn('didFocusSubscription');
        let i = 0;
        if (i == 1)
            mockFn();
        const tree = renderer.create(<FavoritesScreen />).getInstance();
        setImmediate(() => {
            tree.componentWillUnmount();
            expect(tree.state.isCancelled).toEqual(true);
        });
    });

    it('renders list events', () => {
        const mockFn = jest.fn('didFocusSubscription');
        let i = 0;
        if (i == 1)
            mockFn();
        mockAxios.onGet('http://localhost:3030/events/favorites?').reply(200, {
            count: 3,
            events: [{
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP'
            },
            {
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP'
            },
            {
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP'
            }]
        });

        const events = renderer.create(
            <FavoritesScreen
                loading={true}
            />
        );
        let tree = events.toJSON();
        expect(tree).toMatchSnapshot();
    });

});

describe('Get events', () => {
    it('Get events from API', async() => {

        mockAxios.onGet().reply(200,
            [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                favorite: [1]
            }]
        );

        const navigation = { getParam: jest.fn() };

        const wrapper = await shallow(<FavoritesScreen
            loading={true}
            navigation={navigation}
        />);

        await wrapper.instance().getFavoriteEvents();

        expect(wrapper.state().events.length).toEqual(3);
        expect(wrapper.state().events[2].is_favorite).toEqual(true);

    });
});

describe('Favorites', () => {
    it('Marks event as favorite - success', async() => {

        mockAxios.onPost('http://' + global.api + ':3030/favorite').reply(200);

        jest.mock('expo', () => ({
            SecureStore: {
                getItemAsync: jest.fn(() => { return 'asc'; })
            }
        }));

        const wrapper = shallow(<FavoritesScreen
            loading={true}
            category={null}
            entity={null}
        />);

        await wrapper.setState({
            events: [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            }]
        });

        await wrapper.instance().onFavorite(1);

        expect(wrapper.state().events[0].is_favorite).toEqual(false);
    });

    it('Marks event as favorite - error', () => {

        mockAxios.onPost('http://' + global.api + ':3030/favorite').reply(400);

        jest.mock('expo', () => ({
            SecureStore: {
                getItemAsync: jest.fn(() => { return 'asc'; })
            }
        }));

        const wrapper = shallow(<FavoritesScreen
            loading={true}
        />);

        wrapper.setState({
            events: [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            }]
        });

        wrapper.instance().onFavorite(1);

        expect(wrapper.state().events[0].is_favorite).toEqual(true);
    });
});

describe('Infinite scroll', () => {
    it('Handles scroll', async() => {

        mockAxios.onGet().reply(200,
            [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                favorite: [1]
            }]
        );

        const navigation = { getParam: jest.fn() };

        const wrapper = await shallow(<FavoritesScreen
            loading={true}
            navigation={navigation}
        />);

        await wrapper.setState({ updateCall: true });
        await wrapper.instance().getFavoriteEvents();

        expect(wrapper.state().events.length).toEqual(3);

        const event = {
            isTrusted: true,
            type: "scroll",
            nativeEvent: {
                contentOffset: {
                    y: 1000
                }
            }
        };

        await wrapper.instance().handleScroll(event);

        expect(wrapper.state().events.length).toBeGreaterThanOrEqual(3);

    });

    it('Get more events on scroll handling', async() => {

        mockAxios.onGet().reply(200,
            [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                favorite: [1]
            }]
        );

        const navigation = { getParam: jest.fn() };

        const wrapper = await shallow(<FavoritesScreen
            loading={true}
            navigation={navigation}
        />);

        await wrapper.instance().getFavoriteEvents();

        expect(wrapper.state().events.length).toEqual(3);

        let page = wrapper.state().eventsPage + 1;

        await wrapper.setState({ eventsPage: page });

        await wrapper.instance().getFavoriteEvents();

        expect(wrapper.state().events.length).toBeGreaterThanOrEqual(3);
    });
});

describe('Pull refresh', () => {
    it('Refreshes events on pull', async() => {

        mockAxios.onGet().reply(200,
            [{
                id: 1,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: true
            },
            {
                id: 2,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                is_favorite: false
            },
            {
                id: 3,
                title: 'Title',
                description: 'description',
                start_date: '2018-10-27 11:11:00',
                end_date: '2018-10-28 11:11:00',
                initials: 'FEUP',
                favorite: [1]
            }]
        );

        const navigation = { getParam: jest.fn() };

        const wrapper = await shallow(<FavoritesScreen
            loading={true}
            category={null}
            entity={null}
            navigation={navigation}
        />);

        await wrapper.instance().getFavoriteEvents();

        expect(wrapper.state().events.length).toEqual(3);

        let page = wrapper.state().eventsPage + 1;

        await wrapper.setState({ eventsPage: page });

        expect(wrapper.state().eventsPage).toEqual(1);

        await wrapper.instance()._onRefresh();

        expect(wrapper.state().eventsPage).toEqual(0);
    });
});
