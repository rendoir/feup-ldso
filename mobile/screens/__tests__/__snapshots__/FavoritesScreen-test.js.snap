// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FavoritesScreen />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <AppLoading />,
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <AppLoading />,
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`App snapshot renders list events 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
  virtual={undefined}
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.4)",
          "flex": 1,
          "justifyContent": "flex-end",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "elevation": 4,
            "height": NaN,
            "maxHeight": 667,
            "minHeight": 56,
            "opacity": 1,
            "padding": 15,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      />
    </View>
  </Modal>
</View>
`;
